### ISSUE TITLE
Support for Batect

### ISSUE THREAD

**Issue by charleskorn** [renovatebot/renovate#7891]
**What would you like Renovate to be able to do?**

It would be great if Renovate had support for [Batect](https://github.com/batect/batect). 

There are three parts to this, each of which could be handled separately:

* updating Batect's wrapper script
* updating image references in configuration files (would be very similar to what is done for Docker Compose)
* updating bundle references in configuration files

**Did you already have any implementation ideas?**

Updating image references in configuration files should be pretty similar to what is done for Docker Compose, so I assume that could serve as a source of inspiration.

Both bundle references and the wrapper script are handled through GitHub releases, so we should be able to reuse the existing datasource for that. 

I'm happy to have a go at this myself if this is something you're interested in.

**Comment by rarkins** [comment]
@charleskorn sounds great! It doesn't seem you would need new datasources or versioning, so hopefully it's little more than some extract scripts for a new manager or two. 

Are the references you want to update all within the same file/file type, or within separate types of files? If it's separate like package.json and .nvmrc are then you'd want separate Renovate "manager" entries, but if it's all contained within the same file format then a single manager would be best. 

Feel free to email me for an invite to our dev Slack channel if you'd like any advice via chat.

**Comment by charleskorn** [comment]
> Are the references you want to update all within the same file/file type, or within separate types of files?

Image and bundle references are in the same file type (a YAML file normally called `batect.yml`), but the wrapper script is two separate files (`batect` and `batect.cmd`, similar to Gradle's wrapper). So it sounds like I should create two manager entries - something like `batect-config` and `batect-wrapper`?

**Comment by rarkins** [comment]
I think call the config one `batect` and then the wrapper could be `batect-wrapper`. Best implemented in separate PRs.

**Comment by charleskorn** [comment]
The third and final part of this is adding support for updating bundles. These are references to Git repositories, and the format looks like this: 

```yaml
# rest of batect.yml configuration file

include:
  - type: git
    repo: https://github.com/batect/shellcheck-bundle.git
    ref: 0.1.0

  - type: git
    repo: https://github.com/batect/hadolint-bundle.git
    ref: 0.2.0

  - type: git
    repo: https://github.com/batect/another-bundle.git
    ref: 0.2.0
```

Extracting dependencies from this format isn't difficult, but correctly updating them is - the default behaviour of doing a search for the old version and replacing it with the new version could easily update the wrong dependency. For example, if we were trying to update `another-bundle` from `0.2.0` to `0.3.0`, with the standard behaviour, `hadolint-bundle` would be updated instead because its `0.2.0` would be matched and replaced first.

Are there any examples of other managers that have to deal with this problem? Or do you have any suggestions on how to tackle this? Ideally we'd be able to parse the YAML to an AST and update the correct entry in the tree, which would preserve formatting, comments etc, but I can't find any examples of that.

**Comment by rarkins** [comment]
I suggest you test it and it will magically work :)

The autoReplace function is followed by an extract of the new content to verify that the extracted deps match what we expect, so any wrong replace will fail and be followed by a second autoReplace attempt.

**Comment by charleskorn** [comment]
Magic :) That should be very easy to implement then.

**Comment by charleskorn** [comment]
Implemented with #7898, #7939 and #7977.
